core/apple/test/SourceNSInputStreamTest.kt

```diff
@@ -0,0 +1,354 @@
+        val buffer = Buffer()
+        buffer.writeString("abc")
+        val source = RealSource(buffer)
+        assertFalse(source.closed)
+
+        val input = source.asNSInputStream()
+        input.open()
+        input.close()
+        assertTrue(source.closed)
+        assertEquals(NSStreamStatusClosed, input.streamStatus)
+
+        val byteArray = ByteArray(4)
+        byteArray.usePinned {
+            val cPtr = it.addressOf(0).reinterpret<uint8_tVar>()
+
+            byteArray.fill(-5)
+            assertEquals(-1, input.read(cPtr, 4U))
+            assertNotNull(input.streamError)
+            assertEquals("Underlying source is closed.", input.streamError?.localizedDescription)
+            assertEquals("[-5, -5, -5, -5]", byteArray.contentToString())
+        }
+    }
+
+    @Test
+    fun delegateTest() {
+        val runLoop = startRunLoop()
+
+        fun consumeWithDelegate(input: NSInputStream, data: String) {
+            println("consumeWithDelegate(${input::class}, $data)")
+            val opened = Mutex(true)
+            val read = atomic(0)
+            val completed = Mutex(true)
+
+            input.delegate = object : NSObject(), NSStreamDelegateProtocol {
+                val sink = ByteArray(data.length)
+                override fun stream(aStream: NSStream, handleEvent: NSStreamEvent) {
+                    println("${input::class} $data event = ${handleEvent.asString()} (${NSThread.currentThread.name})")
+                    assertEquals("run-loop", NSThread.currentThread.name)
+                    when (handleEvent) {
+                        NSStreamEventOpenCompleted -> opened.unlock()
+                        NSStreamEventHasBytesAvailable -> {
+                            sink.usePinned {
+                                assertEquals(1, input.read(it.addressOf(read.value).reinterpret(), 1U))
+                                read.value++
+                            }
+                        }
+                        NSStreamEventEndEncountered -> {
+                            assertEquals(data, sink.decodeToString())
+                            input.close()
+                            completed.unlock()
+                        }
+                    }
+                }
+            }
+            input.scheduleInRunLoop(runLoop, NSDefaultRunLoopMode)
+            input.open()
+            runBlocking {
+                opened.lockWithTimeout()
+                completed.lockWithTimeout()
+            }
+            assertEquals(data.length, read.value)
+        }
+
+        consumeWithDelegate(NSInputStream(data = "default".encodeToByteArray().toNSData()), "default")
+        consumeWithDelegate(Buffer().apply { writeString("custom") }.asNSInputStream(), "custom")
+        consumeWithDelegate(NSInputStream(data = NSData.data()), "")
+        consumeWithDelegate(Buffer().asNSInputStream(), "")
+        CFRunLoopStop(runLoop.getCFRunLoop())
+    }
+
+    @Test
+    fun uploadTest() {
```

Comment:
It was a good test for verifying a real world use case, but probably not necessary since we've covered the delegate lifecycle in other tests. I'll just remove it.